src/Bicep.IO/Abstraction/ResourceIdentifier.cs

```diff
@@ -0,0 +1,116 @@
+
+            this.Scheme = scheme;
+            this.Authority = authority;
+            this.Path = CanonicalizePath(path);
+        }
+
+        public string Scheme { get; }
+
+        public string? Authority { get; }
+
+        public string Path { get; }
+
+        public bool IsFile => this.Scheme.Equals("file", StringComparison.OrdinalIgnoreCase);
+
+        public bool IsLocal => string.IsNullOrEmpty(this.Authority) || this.Authority.Equals("localhost", StringComparison.OrdinalIgnoreCase);
+
+        public static implicit operator string(ResourceIdentifier identifier) => identifier.ToString();
+
+        public override string ToString() => this.TryGetLocalFilePath() ?? this.ToUriString();
+
+        // See: The "file" URI Scheme (https://datatracker.ietf.org/doc/html/rfc8089).
+        // Note that we don't handle the case where the host IP resolves to the local machine.
+        public string? TryGetLocalFilePath() => this.IsFile && this.IsLocal ? new UriBuilder { Scheme = Scheme, Host = "", Path = Path }.Uri.LocalPath : null;
+
+        // See: Uniform Resource Identifier (URI): Generic Syntax (https://datatracker.ietf.org/doc/html/rfc3986).
+        public string ToUriString() => this.Authority is null ? $"{Scheme}:{Path}" : $"{Scheme}://{Authority}{Path}";
+
+        private static string CanonicalizePath(string path)
+        {
+            var hasTrailingSlash = path.EndsWith('/');
+            var segments = path.Split('/', StringSplitOptions.RemoveEmptyEntries);
+            var stack = new Stack<string>();
+
+            foreach (var segment in segments)
+            {
+                if (segment == ".")
+                {
+                    continue;
+                }
+
+                if (segment == "..")
+                {
+                    if (stack.Count > 0)
+                    {
+                        stack.Pop();
+                    }
+                }
+                else
+                {
+                    stack.Push(segment);
+                }
+            }
+
+            var canonicalPath = string.Join("/", stack.Reverse());
+
+            return hasTrailingSlash
+                ? "/" + canonicalPath + "/"
+                : "/" + canonicalPath;
+        }
+
+        public override int GetHashCode()
+        {
+            var hash = new HashCode();
+
+            hash.Add(Scheme, StringComparer.Ordinal);
+            hash.Add(Authority, StringComparer.Ordinal);
+            hash.Add(Path, GlobalSettings.PathComparer);
+
+            return hash.ToHashCode();
+        }
+
+        public override bool Equals(object? @object) => @object is ResourceIdentifier other && this.Equals(other);
+
+        public bool Equals(ResourceIdentifier other) =>
+            string.Equals(Scheme, other.Scheme, StringComparison.Ordinal) &&
+            string.Equals(Authority, other.Authority, StringComparison.Ordinal) &&
+            string.Equals(Path, other.Path, GlobalSettings.PathComparison);
```

Comment:
Shouldn't the behavior between OSes be the same unless scheme is `file`?

Reply:
Ah good catch. Let me fix it.

Reply:
> I wonder if it's worth making the OS-specific behavior ultra-clear and e.g. making a specific method for this, rather than baking it into the default equality comparer.

It's a long story...but that's what I implemented originally where I put the logic in `Bicep.Core.FileSystem` (this is essentially what [rfc8089 section 2](https://datatracker.ietf.org/doc/html/rfc8089#section-2 suggests). While this design was conceptually better because it avoided leaking abstractions, it ironically ended up complicating the code by a lot. It also made it more difficult to replace System.Uri in Bicep.Core and eliminate PathHelper.PathComparer.

Additionally, I later discovered that MockFileSystem accounts for the OS platform, which would make the test code more complex if we followed this approach.

This current solution is definitely not perfect, but it's a compromise that simplifies the implementation. I'd be open to revisiting the design once the entire migration is complete. For example, there might be an opportunity to get it right if `BicepSourceFile` doesn’t hold a direct reference to `ResourceIdentifier` but instead uses a `IFileHandle`.